[v18] AWS EC2 Discovery: support multiple accounts under Organization

marcoandredinis opened the pull request:
Backport #61940, #62339, #62022, #62023, #62395, #62405, #62964 to v18

Changelog: Added support for discovering EC2 instances in all the accounts under an AWS Organization.

Changelog: Added support for EC2 instances to join based on their AWS Organization.

AWS SDK for Organizations API was updated to the oldest version that supports the State field: 1.45.0
Previous versions, including the one that was being used (1.38.x) do not support the State field, and instead support the Status field.
We could use the Status field, but that is going to be removed in 2026/September: https://aws.amazon.com/about-aws/whats-new/2025/09/aws-organizations-provides-account-state-information-member-accounts/
V18 must be supported until 2027/August: https://goteleport.com/docs/upcoming-releases/#supported-releases

So, even if migrated the code to use the status field, by September, 2026, we would start considering non-active accounts.

Upgrading to 1.45.0 should be fine. No breaking change was introduced and no other package needed to be updated.

## Test Plan

### Self-hosted:

The tests will be done using what's described in this guide:
https://goteleport.com/docs/ver/19.x/enroll-resources/auto-discovery/servers/ec2-discovery/ec2-discovery-organization/

- [x] Org wide EC2 Discovery
<img width="1440" height="1296" alt="image" src="https://github.com/user-attachments/assets/6d952970-e2f6-48f3-9e07-b6c37094d594" />

- [x] IAM Join using organization id
<img width="1494" height="1264" alt="image" src="https://github.com/user-attachments/assets/057e1e11-360f-4eac-b9ad-7dcc5342bb75" />

<img width="1042" height="618" alt="image" src="https://github.com/user-attachments/assets/2d3f65c6-5c0a-447f-900b-95b26380fdb3" />


### Teleport Cloud w/ integration
Slightly change on the guide above, this time using aws oidc integration to allow access to aws APIs.
- [x] Org wide EC2 Discovery
<img width="596" height="740" alt="image" src="https://github.com/user-attachments/assets/5a1698c7-5061-4186-a221-e559d64d058d" />

- [x] IAM Join using organization id
<img width="1474" height="1472" alt="image" src="https://github.com/user-attachments/assets/ac16a4eb-b5c0-47d7-9360-096f38481120" />
<img width="484" height="322" alt="image" src="https://github.com/user-attachments/assets/4b8e8140-641c-42c7-a9a2-380f96cb5aa3" />

marcoandredinis committed:
IAM Provision Token: allow joins by AWS Organization (proto) (#62022)

marcoandredinis committed:
AWS EC2 Discovery: support multiple accounts under Organization (#61940)

* AWS EC2 Discovery: support multiple accounts under Organization

* fix loop

* review

* move assume role to role name

* move assume role to role name

* remove wildcard from exclude possible values

* remove list all accounts api call

marcoandredinis committed:
update aws sdk organizations pkg to 1.45.0

only this version supports the state field
we could use the status field, which is similar but that field will be
retired in 2026/Sep/9

it didn't update any other dependency

marcoandredinis committed:
Allow IAM Joins to use an Organization ID with integration creds (#62395)

This PR adds a new field in the provision token, so that users can
define an integration whose credentials will be used to retreive the
organization id for a given account.

This will allow the Auth Server to call the AWS API
organizations:DescribeAccount using the integration credentials.
Previously, only self-hosted clusters could allow iam joins by
organization (by providing ambient credentials to the Auth Server)

marcoandredinis committed:
Allow nodes to join the cluster based on their AWS Organization (#62023)

* Allow nodes to join the cluster based on their AWS Organization

* remove unused struct field

* add cache to the DescribeAccount API response

* review pt1

* review pt2

* remove setter

marcoandredinis committed:
Docs: EC2 discovery in multiple accounts under an AWS Organization (#62339)

* rework as a dedicated guide

* add mention of orgs guide in single account guides

marcoandredinis committed:
IAM Joins by organization with support for integration credentials (#62405)

marcoandredinis committed:
Docs: fix EC2 Discovery Config install key names (#62964)

rosstimothy reviewed (dismissed):
Can you please add a manual test plan and confirm that all tests are passing?

marcoandredinis committed:
Release v18.7.0-dev.marco.1

marcoandredinis committed:
Docs: fix tags key in EC2 Discovery guides (#63730)

Fixed indentation and moved it near the regions in all examples.

Filtering by tag, IMO, is a similar concern to filtering by region, so
having them together makes sense.
